### PR TITLE
WIP: test: add stress test for ExecSyncRequests

### PR DESCRIPTION
**Pull request by haircommander**
Note: it sometimes takes a couple of runs to reproduce

Signed-off-by: Peter Hunt <pehunt@redhat.com>

### CODE REVIEW

File: pkg/client/client_test.go [hunk 1]
```diff
@@ -374,6 +375,51 @@ var _ = Describe("ConmonClient", func() {
 				Expect(result).NotTo(BeNil())
 				Expect(result.TimedOut).To(Equal(true))
 			})
+
+			testName = "should handle many requests"
+			if terminal {
+				testName += " with terminal"
+			}
+			FIt(testName, func() {
```
Comment (haircommander) [inline pkg/client/client_test.go:383]: WIP because of this line. I found it easier to debug with it

**Comment by haircommander** [comment]
https://github.com/containers/conmon-rs/pull/300 also doesn't yet fix this problem :(

**Comment by saschagrunert** [comment]
How does that work together with https://github.com/containers/conmon-rs/pull/300?

**Comment by saschagrunert** [comment]
So, disconnecting the stdio streams in non terminal situations may be the problem here since it is globally applied. Does this work?

1. Crewte container with terminal
2. Exec sync without terminal
3. Exec sync with terminal

File: pkg/client/client_test.go [hunk 2]
```diff
@@ -374,6 +375,51 @@ var _ = Describe("ConmonClient", func() {
...
+
+				// Wait for container to be running
+				Eventually(func() error {
+					return rr.RunCommandCheckOutput("running", "list")
+				}, time.Second*10).Should(BeNil())
+
+				var wg sync.WaitGroup
+				for i := 0; i < 15; i++ {
+					go func() {
+						wg.Add(1)
```
Comment (rphillips) [inline pkg/client/client_test.go:409]: the wg.Add should be prior to the go func since it'll race.

**Comment by haircommander** [comment]
> Does it also break when `terminal == true`?
> 
> For non terminal scenarios, disconnecting the stdio streams may be the problem here since it is globally applied. We probably should do the setup only once.

terminal always succeeds! I'm not sure about disconnecting. we can do about 3 or 4 reliably, once it gets higher it usually stalls though

File: pkg/client/client_test.go (hunk 2 above)
Comment (haircommander) [reply pkg/client/client_test.go:409]: ah good to know

**Comment by saschagrunert** [comment]
We still use the single threaded runtime so it may be possible that we deadlock at some point:

https://github.com/containers/conmon-rs/blob/937f43846f297670feb741fa1998f579d70bcb49/conmon-rs/server/src/server.rs#L83

If changing the runtime to multi threaded solve the problem, then we would know that it’s indeed a thread dead lock.

**Comment by haircommander** [comment]
> 

Good theory! Interestingly enough, we still hit this error if we set to multi_thread

**Comment by rphillips** [comment]
Looks like we are deadlocking when terminal == false

**Comment by haircommander** [comment]
> We are reusing a file descriptor... maybe we are not clearing a handle on it somewhere...

that would be for the terminal case, so I don't think it'd be the cause here. Unfortunately the non terminal conmonrs logs aren't printed to stdout because of the way it currently works. Also, FYI I believe that's both stdout and stderr streams being registered. The way the OCI runtime specifies terminal requires this to be the case

**Comment by saschagrunert** [comment]
@haircommander the logs are still accessible via the journal

**Pull request closed without merging**